### PR TITLE
Test DDNS connectivity when connecting to a zone

### PR DESCRIPTION
**Pull request by mitruly**
When "creating" (ie. connecting to) a zone in VinylDNS, perform a DDNS delete and add for a test record set to check connectivity.
- Record set is named `vinyldns-ddns-connectivity-test` of type `TXT`. If a user is connecting to a zone with an existing record, its record data will be replaced.
- Add pre-requisite for delete messages, asserting that record to delete actually exists (so DNS response is `NXRRSET`).
- When attempting to delete the record, `RecordSetNotFound` (`NXRRSET`) and `NameNotFound` (`NXDOMAIN`) errors are suppressed.
- Update TSIG key to HMAC-SHA512 algorithm (from default of HMAC-MD5).

Updated main source files, unit and func tests as well as API documentation.

### CODE REVIEW

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 1]
```diff
@@ -71,6 +71,41 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection, scheduler: Sche
...
+      "vinyldns-test",
+      RecordType.TXT,
+      30,
+      RecordSetStatus.Pending,
+      DateTime.now,
+      records = List(TXTData("connection test")))
+
+    def performDdnsChecks(rsList: List[RecordSet]): Result[Unit] =
+      if (rsList.isEmpty) {
+        // Perform add then delete of test record
```
Comment (pauljamescleary) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: I think you should change the check up.

The logic currently follows:
1. If the test record does not exist, then add it and then delete it
2. If the test record does exist, then delete it and then add it

You will end up in two different places.   If the record does not exist (first case), it will not exist after the logic.   If the record does exist (second case), it will still exist after the logic.

Rather than doing the resolve, I would recommend:

1. Attempt to delete the record.  If you get a not found error coming back, that is ok!  Keep moving, swallowing that error
2. Create the record

I think doing an immediate delete is not what we want, but open to discussion.  DNS servers maybe eventually consistent, so that churn may wind up with some unexpected errors?  Perhaps just leave it there after step 2. 

**Note: we do not need to verify the record was created.  If we get back a NoError on the add we should be good**.

Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: @pauljamescleary Thanks for the comment - discussion would definitely be useful for me.
- One caveat that I found without using `resolve` is that if we are deleting a record set that matches the same name and type, we would potentially be re-adding it with different record data, TTL and other settings. I was playing around with `resolve` and I think that `NoError` is also returned in the DNS response for a delete of a non-existent record.
- It might also be confusing if we never delete the record since it would show up in the zone unless we are somehow filtering that out.

With that said, if we are fine with saying that `vinyldns-test` is off-limits to users and will be created in a zone upon connection, then I think a simple delete and add like you suggest would suffice.

File: docker/bind9/etc/named.conf.local [hunk 2]
```diff
@@ -11,6 +11,11 @@ key "vinyldns." {
   secret "nzisn+4G2ldMn0q1CV3vsg==";
 };
 
+key "vinyldns-no-updates" {
```
Comment (mitruly) [inline docker/bind9/etc/named.conf.local]: Generated `secret` using `ddns-confgen`. `secret` must match algorithm; otherwise, invalid TSIG key error is thrown.

File: docker/bind9/zones/one-time.hosts [hunk 3]
```diff
@@ -12,3 +12,4 @@ test             IN  A   3.3.3.3
 test             IN  A   4.4.4.4
 @                IN  A   5.5.5.5
 already-exists   IN  A   6.6.6.6
+vinyldns-ddns-connectivity-test    IN  TXT "random existing text"
```
Comment (mitruly) [inline docker/bind9/zones/one-time.hosts]: Testing replacement of existing record matching DDNS test record with different `text`.

File: modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py [hunk 4]
```diff
@@ -106,11 +106,11 @@ def test_list_recordset_changes_paging(zone_history_context):
     response_2 = client.list_recordset_changes(original_zone['id'], start_from=response_1['nextId'], max_items=3)
     # nextId differs local/in dev where we get exactly the last item
     # Requesting one over the total in the local in memory dynamo will force consistent behavior.
-    response_3 = client.list_recordset_changes(original_zone['id'], start_from=response_2['nextId'], max_items=11)
```
Comment (mitruly) [inline modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py]: List record set change tests needed to get updated to include test DDNS record.

File: modules/api/functional_test/live_tests/zones/create_zone_test.py [hunk 5]
```diff
@@ -33,9 +33,12 @@
      'records': [{u'address': u'5.5.5.5'}]},
     {'name': u'already-exists',
      'type': u'A',
-     'records': [{u'address': u'6.6.6.6'}]}]
+     'records': [{u'address': u'6.6.6.6'}]},
+    {'name': u'vinyldns-ddns-connectivity-test',
```
Comment (mitruly) [inline modules/api/functional_test/live_tests/zones/create_zone_test.py]: Adding check to create zone test.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 6]
```diff
@@ -71,6 +71,26 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection, scheduler: Sche
     }
   }.toResult
 
+  def testDdnsConnectivity(dnsConnection: DnsConnection, zone: Zone): Result[Unit] = {
+    val rs = RecordSet(
```
Comment (mitruly) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: Couldn't find a better place to store this since it is based off of `zone`, which gets passed in.

**Comment by mitruly** [comment]
CodeCov doesn't appear to consistently update with subsequent commits.. This is from running `sbt verify` locally in this branch:
```
[info] Aggregation complete. Coverage was [94.20]
[info] All done. Coverage was [94.20%]
```

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 7]
```diff
@@ -71,6 +71,26 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection, scheduler: Sche
...
+      zone.id,
+      "vinyldns-ddns-connectivity-test",
+      RecordType.TXT,
+      30,
```
Comment (mitruly) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: Was considering changing `TTL` to the max value - would like some thoughts on this.

File: modules/api/functional_test/vinyldns_context.py [hunk 8]
```diff
@@ -5,6 +5,8 @@ class VinylDNSTestContext:
     dns_rev_v6_zone_name = '1.9.e.f.c.c.7.2.9.6.d.f.ip6.arpa.'
     dns_key_name = 'vinyldns.'
     dns_key = 'nzisn+4G2ldMn0q1CV3vsg=='
+    dns_no_updates_key_name = 'vinyldns-no-updates'
+    dns_no_updates_key = 'BjI47RjxzSQyOktSnmbD4A=='
```
Comment (pauljamescleary) [inline modules/api/functional_test/vinyldns_context.py]: this will break deployment as our deployment relies on these items being on our test bind9 servers.

I think our test servers do not support 128 bit keys.  

Ideally, we would use the same key strength as our test bind9 servers for this key (this key is 128 bit).  If you can make it work, that is ideal.

Otherwise, we need to make the key value a parameter to the func tests, passed into conftest.py

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala (hunk 7 above)
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: TTL should be higher for sure, a day would be sufficient.  Noone will really query for it tbh.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 9]
```diff
@@ -71,6 +71,26 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection, scheduler: Sche
...
+      RecordSetStatus.Pending,
+      DateTime.now,
+      records = List(TXTData("connection test")))
+
+    val result = for {
+      _ <- dnsConnection.applyChange(RecordSetChange.forDelete(rs, zone))
```
Comment (pauljamescleary) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: applyChange will fail `Left` with `RecordSetNotFound` or `NameNotFound`.

You need to "swallow" these two cases and proceed to the add.

File: modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneConnectionValidatorSpec.scala [hunk 10]
```diff
@@ -138,6 +144,20 @@ class ZoneConnectionValidatorSpec
       result should be(right)
     }
 
+    "respond with a failure if DDNS connectivity fails" in {
```
Comment (pauljamescleary) [inline modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneConnectionValidatorSpec.scala]: need to add unit tests here:

1. delete returns a `NameNotFound`
1. delete returns a `RecordSetNotFound`
1. delete returns an unexpected error code, like `Refused` (which we see a lot)

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 11]
```diff
@@ -71,6 +71,26 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection, scheduler: Sche
...
 
+  def testDdnsConnectivity(dnsConnection: DnsConnection, zone: Zone): Result[Unit] = {
+    val rs = RecordSet(
+      zone.id,
+      "vinyldns-ddns-connectivity-test",
+      RecordType.TXT,
+      30,
+      RecordSetStatus.Pending,
+      DateTime.now,
+      records = List(TXTData("connection test")))
```
Comment (pauljamescleary) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: rather than connection test, let's put something more meaningful to make sure the value is updated.  Something like the current date and time

File: modules/api/functional_test/vinyldns_context.py (hunk 8 above)
Comment (mitruly) [reply modules/api/functional_test/vinyldns_context.py]: @pauljamescleary I think you are right. I tried using a different strength key, and that failed a few integration tests with some TSIG not authorized error message. I added options to the `conftest.py`.

File: docker/bind9/etc/named.conf.local [hunk 12]
```diff
@@ -7,8 +7,13 @@
 //include "/etc/bind/zones.rfc1918";
 
 key "vinyldns." {
-  algorithm hmac-md5;
-  secret "nzisn+4G2ldMn0q1CV3vsg==";
+    algorithm hmac-md5;
+    secret "H3qEuLC+7sG6exm7+F+dBwuV+pbu6cJIs0Yq3E607A4dB8yEtk5j5EeP U1HnR6jUFa2g+hYNYW+iubB8Ii90Eg==";
```
Comment (rebstar6) [inline docker/bind9/etc/named.conf.local]: if were changing this here do we not need to change the default connection keys in docker.conf to be this too (?)

File: docker/bind9/zones/one-time.hosts (hunk 3 above)
Comment (rebstar6) [reply docker/bind9/zones/one-time.hosts]: isnt the point of one-time that we only have one test using it?

id probably use a different zone for this check

File: modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py (hunk 4 above)
Comment (rebstar6) [reply modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py]: no, we shouldnt be storing the change in the database i thought (?)

this shouldnt be visible to people

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 13]
```diff
@@ -67,6 +69,31 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection)
     }
   }.toResult
 
+  def testDdnsConnectivity(dnsConnection: DnsConnection, zone: Zone): Result[Unit] = {
+    val rs = RecordSet(
+      zone.id,
+      "vinyldns-ddns-connectivity-test",
+      RecordType.TXT,
+      86400,
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: id think we'd want a low TTL on this guy so it doesnt hang around

File: modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py (hunk 4 above)
Comment (mitruly) [reply modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py]: @rebstar6 That is the approach that we initially took with this. There was some follow-up conversation at https://github.com/vinyldns/vinyldns/pull/140#discussion_r212969853 where it was suggested to keep this connectivity test record. If we're keeping the record but not updating the database (which might be what you are suggesting), then I will have to update how it is being handled.

File: modules/api/functional_test/live_tests/recordsets/list_recordsets_test.py [hunk 14]
```diff
@@ -54,6 +55,19 @@ def __init__(self, shared_zone_test_context):
         for i in range(7):
             self.all_records[i + 10] = existing_records[i]
 
+        # Adding in DDNS connectivity record
+        self.all_records[17] = {
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/recordsets/list_recordsets_test.py]: same here, if were removing this record, why are people seeing it in listRecords?

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 15]
```diff
@@ -67,6 +69,31 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection)
...
+        .applyChange(RecordSetChange.forDelete(rs, zone))
+        .map(_ => ())
+        .recover {
+          case _: RecordSetNotFound | _: NameNotFound => ()
+        }
+      _ <- dnsConnection.applyChange(RecordSetChange.forAdd(rs, zone))
+    } yield ()
+
+    result.leftMap {
+      case e => ConnectionFailed(zone, s"Unable to test DDNS connectivity to zone: ${e.getMessage}")
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: i dont think "Unable to test DDNS connectivity" is the right message here...Unable to apply changes in zone, or something

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala (hunk 13 above)
Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: Initially this was set to the minimum TTL of `30`, but it was suggested to be bumped up to 24 hours if the record is sitting around.

File: modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneConnectionValidatorSpec.scala [hunk 16]
```diff
@@ -76,6 +78,12 @@ class ZoneConnectionValidatorSpec
     override val opTimeout: FiniteDuration = 10.milliseconds
     override def dnsConnection(conn: ZoneConnection): DnsConnection = testDnsConnection(conn)
     override def loadDns(zone: Zone): IO[ZoneView] = testLoadDns(zone)
+    override def testDdnsConnectivity(dnsConnection: DnsConnection, zone: Zone): Result[Unit] =
+      if (zone == testZone) {
+        ().toResult
+      } else {
+        Left(ConnectionFailed(zone, s"DDNS connection failed").asInstanceOf[Throwable]).toResult
```
Comment (rebstar6) [inline modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneConnectionValidatorSpec.scala]: shouldnt need asInstanceOf here

File: modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneConnectionValidatorSpec.scala [hunk 17]
```diff
@@ -130,11 +138,68 @@ class ZoneConnectionValidatorSpec
...
         .resolve(testZone.name, testZone.name, RecordType.SOA)
+      doReturn(NoError(mock[Message]).toResult)
+        .when(mockDnsConnection)
+        .applyChange(any[RecordSetChange])
+
+      val result = awaitResultOf(underTest.validateZoneConnections(testZone).value)
+      result should be(right)
+    }
+
+    "respond with a success if DDNS connectivity test returns NXDOMAIN on delete" in {
```
Comment (rebstar6) [inline modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneConnectionValidatorSpec.scala]: these unit tests dont all have to be on validateZoneConnections, it makes the mocking hard to follow. test testDdnsConnectivity individually and then just have one test confirming that if testDdnsConnectivity fails, validateZoneConnections fails

File: docker/bind9/etc/named.conf.local [hunk 18]
```diff
@@ -7,8 +7,13 @@
 //include "/etc/bind/zones.rfc1918";
 
 key "vinyldns." {
-  algorithm hmac-md5;
-  secret "nzisn+4G2ldMn0q1CV3vsg==";
+	algorithm hmac-sha512;
+	secret "wCZZS9lyRr77+jqfnkZ/92L9fD5ilmfrG0sslc3mgmTFsF1fRgmtJ0rj RkFITt8VHQ37wvM/nI9MAIWXYTvMqg==";
```
Comment (mitruly) [inline docker/bind9/etc/named.conf.local]: To generate this TSIG key, I used: `dnssec-keygen -a HMAC-SHA512 -b 512 -n user vinyldns.`. This *is different* from not specifying `-n user` (since it will default to `-n host`). `dnssec-keygen` can be installed through bind-utils.

`ddns-confgen` invokes `tsig-keygen`. Not sure if `ddns-confgen` can be used to generate the same exact key, but `tsig-keygen` might.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala [hunk 19]
```diff
@@ -35,7 +35,7 @@ object DnsZoneViewLoader extends DnsConversions {
   def dnsZoneTransfer(zone: Zone): ZoneTransferIn = {
     val conn =
       zone.transferConnection.getOrElse(VinylDNSConfig.defaultTransferConnection).decrypted()
-    val TSIGKey = new TSIG(conn.keyName, conn.key)
+    val TSIGKey = new TSIG(TSIG.HMAC_SHA512, conn.keyName, conn.key)
```
Comment (mitruly) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: If we are switching to HMAC-SHA512 in this PR, this is more of a stop gap since different keys can have different algorithms. Need to make this more configurable, but don't want to blow up this PR.

File: modules/api/functional_test/conftest.py [hunk 20]
```diff
@@ -19,8 +19,14 @@ def pytest_addoption(parser):
                      help="The zone name that will be used for testing")
     parser.addoption("--dns-key-name", dest="dns_key_name", action="store", default="vinyldns.",
                      help="The name of the key used to sign updates for the zone")
-    parser.addoption("--dns-key", dest="dns_key", action="store", default="nzisn+4G2ldMn0q1CV3vsg==",
-                     help="The tsig key")
+    parser.addoption("--dns-key", dest="dns_key", action="store",
+                    default="wCZZS9lyRr77+jqfnkZ/92L9fD5ilmfrG0sslc3mgmTFsF1fRgmtJ0rj RkFITt8VHQ37wvM/nI9MAIWXYTvMqg==",
```
Comment (pauljamescleary) [inline modules/api/functional_test/conftest.py]: do we still need this?  in other words, can't we use the new key in our internal vinyldns bind9 servers?

If we can use the same key, we can drop this all together, right?

File: modules/api/src/it/scala/vinyldns/api/domain/dns/DnsConversionsIntegrationSpec.scala [hunk 21]
```diff
@@ -34,10 +34,18 @@ class DnsConversionsIntegrationSpec
...
     "test@test.com",
     ZoneStatus.Active,
-    connection =
-      Some(ZoneConnection("vinyldns.", "vinyldns.", "nzisn+4G2ldMn0q1CV3vsg==", "127.0.0.1:19001")),
-    transferConnection =
-      Some(ZoneConnection("vinyldns.", "vinyldns.", "nzisn+4G2ldMn0q1CV3vsg==", "127.0.0.1:19001"))
+    connection = Some(
+      ZoneConnection(
+        "vinyldns.",
+        "vinyldns.",
```
Comment (pauljamescleary) [inline modules/api/src/it/scala/vinyldns/api/domain/dns/DnsConversionsIntegrationSpec.scala]: I am ok with replacing the existing "test" key, but we just need to be sure that we haven't broken things elsewhere (like our internal deployment)

File: modules/api/src/main/scala/vinyldns/api/domain/dns/DnsConversions.scala [hunk 22]
```diff
@@ -511,14 +512,18 @@ trait DnsConversions {
...
+        update.delete(r)
+      }
       addRecords.foreach(update.add)
     }
     Right(update)
   }
 
   def toDeleteRecordMessage(r: DNS.RRset, zoneName: String): Either[Throwable, DNS.Update] = {
     val update = new DNS.Update(zoneDnsName(zoneName))
+    update.present(r.getName, r.getType)
```
[the remark this answers is not in the document]
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/dns/DnsConversions.scala]: Not sure if we should add this condition here.  In the `RecordSetChangeHandler` during the apply, the only thing we check for is `NoError`.  In reality, getting a `NXRRSET` back during a delete needs to be properly handled depending on context.  For the most part, we respect idempotency, so deleting a record set that doesn't exist is _ok_, and we should not error our.

It is ok if we don't get `NXRRSET` back in the validation.  As long as we swallow errors appropriately.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala (hunk 13 above)
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: Yea, this should be high.  Otherwise, it will expire in caching resolvers and those resolvers will query upstream to the authoritative masters for updates to the record.

The record set always hangs around with the current implementation, we just want to avoid frequent cache refreshes.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala (hunk 19 above)
Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: Created https://github.com/vinyldns/vinyldns/issues/156 so wrap up configurability of TSIG key algorithms.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 23]
```diff
@@ -71,6 +72,31 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection, scheduler: Sche
...
+      RecordType.TXT,
+      86400,
+      RecordSetStatus.Pending,
+      DateTime.now,
+      records = List(TXTData("connection test")))
+
+    val result = for {
+      _ <- dnsConnection
+        .applyChange(RecordSetChange.forDelete(rs, zone))
+        .map(_ => ())
```
[the remark this answers is not in the document]
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: use `.as(())` instead of `.map(_ => ())`

File: modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py (hunk 4 above)
Comment (rebstar6) [reply modules/api/functional_test/live_tests/recordsets/list_recordset_changes_test.py]: oh, hrmm. so every zone in vinyl will have this record in it after creation (?)

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 24]
```diff
@@ -67,6 +69,31 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection)
...
+        .recover {
+          case _: RecordSetNotFound | _: NameNotFound => ()
+        }
+      _ <- dnsConnection.applyChange(RecordSetChange.forAdd(rs, zone))
```
[the remark this answers is not in the document]
Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: In the case of a bad key, this is failing on the delete first. @pauljamescleary might be able to provide more insight, but I think we wanted to make sure there was some update of some sort.

I suppose a record existing would also run into a bad key error first, so it might suffice.

[the remark this answers is not in the document]
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: It probably makes sense to test the different cases.  What happens if we use a bad key?  What happens if the key is not authorized?

Should be rather simple to test on a docker bind9.

[the remark this answers is not in the document]
Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: We already have separate func tests that test both of these cases.

[the remark this answers is not in the document]
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: I think @mitruly is, do we need to do an `add` at all here?  Can we just issue the delete?  The idea is to verify that the key is valid and has access to make changes in the zone.

[the remark this answers is not in the document]
Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: @pauljamescleary Right, I was discussing offline with @rebstar6 and that makes sense. I'll make the changes accordingly. 👍

File: modules/api/functional_test/vinyldns_context.py [hunk 25]
```diff
@@ -1,16 +1,18 @@
+import os
+
 class VinylDNSTestContext:
     dns_ip = 'localhost'
     dns_zone_name = 'vinyldns.'
     dns_rev_v4_zone_name = '30.172.in-addr.arpa.'
     dns_rev_v6_zone_name = '1.9.e.f.c.c.7.2.9.6.d.f.ip6.arpa.'
-    dns_key_name = 'vinyldns.'
-    dns_key = 'nzisn+4G2ldMn0q1CV3vsg=='
+    dns_key_name = os.getenv('DNS_KEY_NAME', 'vinyldns.')
```
Comment (pauljamescleary) [inline modules/api/functional_test/vinyldns_context.py]: Sorry if I wasn't clear on this.  If the keys are the same in _all_ environments, then we should be able to just hard code them here, as opposed to using environment variables.  Is there a reason that you decided to use environment variables?

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala (hunk 19 above)
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: This will probably break the other place we create TSIG keys (`DnsConnection`).

We cannot assume the key algorithm of the backend.  Until now, we have used the default constructor for TSIG, which assumes an algorithm of HMAC_MD5

That has _always_ worked for us.

I am worried that this will break everything, especially if what we are using today is HMAC_MD5

File: modules/api/functional_test/vinyldns_context.py (hunk 25 above)
Comment (mitruly) [reply modules/api/functional_test/vinyldns_context.py]: @pauljamescleary Changed this to hard code.

**Comment by mitruly** [comment]
@pauljamescleary The `TSIG` algorithm was already changed in `DnsConnection` in https://github.com/vinyldns/vinyldns/pull/140/commits/5ccc17de069c876766562c4a6a317c1e07f5454e#diff-9ccc2444aca86af06e87f0385327bafa; otherwise the tests would not pass.

File: modules/docs/src/main/tut/api/create-zone.md [hunk 26]
```diff
@@ -7,7 +7,14 @@ section: "api"
 # Create Zone
 
 Connects user to an existing zone. User must be a member of the group that has access to the zone. Connection info is optional,
-if no info is provided the default VinylDNS connections will be used
+if no info is provided the default VinylDNS connections will be used. 
+
+When attempting a zone connection, VinylDNS will attempt to perform a DDNS delete of a `TXT` record with name
```
Comment (pauljamescleary) [inline modules/docs/src/main/tut/api/create-zone.md]: is this still happening?

File: modules/api/functional_test/vinyldns_context.py (hunk 25 above)
Comment (mitruly) [reply modules/api/functional_test/vinyldns_context.py]: The commit that I pushed up should have addressed this: https://github.com/vinyldns/vinyldns/pull/140/files#diff-4dc0a0a54250794c10d0d0ecc4f11369

File: modules/docs/src/main/tut/api/create-zone.md (hunk 26 above)
Comment (mitruly) [reply modules/docs/src/main/tut/api/create-zone.md]: Yes, we are attempting a DDNS delete of the record still. If it succeeds, any existing record match will be deleted. I updated the verbiage from delete/add/lingering record.

Comment (pauljamescleary) [reply modules/docs/src/main/tut/api/create-zone.md]: Right, I was looking at an older commit where it said it was adding the record, this is good now

File: modules/api/functional_test/live_tests/zones/create_zone_test.py [hunk 27]
```diff
@@ -7,7 +7,7 @@
 from utils import *
 
 records_in_dns = [
-    {'name': 'one-time.',
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/zones/create_zone_test.py]: what happened to one-time?!

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala [hunk 28]
```diff
@@ -67,6 +69,30 @@ class ZoneConnectionValidator(defaultConnection: ZoneConnection)
...
+      "vinyldns-ddns-connectivity-test",
+      RecordType.TXT,
+      86400,
+      RecordSetStatus.Pending,
+      DateTime.now,
+      records = List(TXTData("connection test")))
+
+    val result = for {
+      _ <- dnsConnection
+        .applyChange(RecordSetChange.forDelete(rs, zone))
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneConnectionValidator.scala]: cool, i like the delete-only

File: modules/api/functional_test/live_tests/zones/create_zone_test.py (hunk 27 above)
Comment (mitruly) [reply modules/api/functional_test/live_tests/zones/create_zone_test.py]: With the change, it would affect all one-time tests so opted to use a different zone
